lib/rack/tempfile_reaper.rb

```diff
@@ -9,25 +9,35 @@ module Rack
   # Ideas/strategy based on posts by Eric Wong and Charles Oliver Nutter
   # https://groups.google.com/forum/#!searchin/rack-devel/temp/rack-devel/brK8eh-MByw/sw61oJJCGRMJ
   class TempfileReaper
+    RESPONSE_FINISHED_HANDLER = ->(env, _, _, _) {
+      env[RACK_TEMPFILES]&.each(&:close!)
+    }
+
     def initialize(app)
       @app = app
     end
 
     def call(env)
       env[RACK_TEMPFILES] ||= []
 
-      begin
-        _, _, body = response = @app.call(env)
-      rescue Exception
-        env[RACK_TEMPFILES]&.each(&:close!)
-        raise
-      end
+      if response_finished = env[RACK_RESPONSE_FINISHED]
+        response_finished << RESPONSE_FINISHED_HANDLER
```

Comment:
```suggestion
        tempfiles = env[RACK_TEMPFILES] ||= []

      if response_finished = env[RACK_RESPONSE_FINISHED]
        response_finished << proc do
          tempfiles.each(&:close!)
        end
```

I'm not in favour of my proposed approach, but I thought I'd mention it.